### PR TITLE
H/3 Server Cert validation callback exception fix

### LINKED ISSUE #55192: [HTTP/3] ServerCertificateCustomValidationCallback causes connection error
Adding `ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator` causes a connection error on send.

Code:
```cs
// HTTP/3
var httpClientHandler = new HttpClientHandler();
httpClientHandler.ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;

using (var client = new HttpClient(httpClientHandler))
{
    var request = new HttpRequestMessage(HttpMethod.Get, $"https://127.0.0.1:{http3Port}/");
    request.Version = HttpVersion.Version30;
    request.VersionPolicy = HttpVersionPolicy.RequestVersionExact;

    // Act
    var response = await client.SendAsync(request).DefaultTimeout();

    // Assert
    response.EnsureSuccessStatusCode();
    Assert.Equal(HttpVersion.Version30, response.Version);
    var responseText = await response.Content.ReadAsStringAsync().DefaultTimeout();
    Assert.Equal("hello, world", responseText);
}
```

Error:
```
    System.Net.Http.HttpRequestException : Connection has been shutdown by transport. Error Code: 0x8041012A (127.0.0.1:5005)
    ---- System.Net.Quic.QuicException : Connection has been shutdown by transport. Error Code: 0x8041012A
```

### PR DESCRIPTION
**Pull request by ManickaP**
Sets and unsets `HttpMessageRequest` to the cert callback when opening new `QuicConnection`.

Fixes #55192

### CODE REVIEW

**Comment by ManickaP** [comment]
I got a quick chat with @wfurt and I'm working on a proper fix, please hold off your review efforts for the moment.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs [hunk 1]
```diff
@@ -104,8 +104,22 @@ private static async ValueTask<SslStream> EstablishSslConnectionAsyncCore(bool a
...
+            if (callback != null && callback.Target is CertificateCallbackMapper mapper)
+            {
+                clientAuthenticationOptions = clientAuthenticationOptions!.ShallowClone(); // Clone as we're about to mutate it and don't want to affect the cached copy
+                Func<HttpRequestMessage, X509Certificate2?, X509Chain?, SslPolicyErrors, bool> localFromHttpClientHandler = mapper.FromHttpClientHandler;
+                clientAuthenticationOptions.RemoteCertificateValidationCallback = (object sender, X509Certificate? certificate, X509Chain? chain, SslPolicyErrors sslPolicyErrors) =>
+                {
+                    bool result = localFromHttpClientHandler(request, certificate as X509Certificate2, chain, sslPolicyErrors);
+                    return result;
+                };
+            }
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:121]: This looks like an almost duplicate of https://github.com/dotnet/runtime/pull/55526/files#diff-9eb150e0af6ce559b3ef3e99949ba3512594fe4c55fbf10e6b739d315899d469R40-R57.  Why are we duplicating the code?  

It also looks like as part of copying/pasting that here, you removed the `localRequest`... why?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs [hunk 2]
```diff
@@ -117,6 +131,17 @@ public static async ValueTask<QuicConnection> ConnectQuicAsync(QuicImplementatio
                 con.Dispose();
                 throw CreateWrappedException(ex, endPoint.Host, endPoint.Port, cancellationToken);
             }
+            finally
+            {
+                if (clientAuthenticationOptions is not null && callback is not null)
+                {
+                    // We cannot unset the request closure variable after the first callaback invocation as we do with SslStream
+                    // since the callback will get triggered multiple times during connection establishment.
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:139]: Ah, is this why you duplicated the code?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs (hunk 1 above)
Comment (ManickaP) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:121]: For msquic, the callback gets called multiple times when the connection is established, so I couldn't unset it in the callback itself, but after the connection is opened.
So I moved the "unsetting" to finally block, where I discard the whole custom callback wrapper and put back the original, thus removing reference to the request.

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:121]: Ok.  Note the localRequest was also there as an optimization.  Right now you're paying for the closure even if there is no callback.

**Comment by wfurt** [comment]
The underlying issue is tracked by https://github.com/microsoft/msquic/issues/1810
Seems platform specific e.g. we get the multiple invocations only on Linux -> Windows work as expected. 

We could put workaround to MsQuicConnection and ignore additional validation callbacks. 
That could be removed when MsQuic is fixed (or we can keep it if it never is) 

Since TLS 13 does not support renegotiation there should never be case when we need to do validation more than once.

**Comment by ManickaP** [comment]
@wfurt will do as you suggest, if feasible. If not, I'll push this through as is.

**Comment by ManickaP** [comment]
In the end, I prevented `MsQuicConnection` to call the cert validation callback multiple times by remembering the first value returned.
Thus, I also removed the code duplication. cc: @stephentoub 
@wfurt could you give it a final review?

**Review by wfurt — APPROVED** [review]
LGTM.

**Pull request merged**